client-libraries/rust/modality-ingest-client/src/types.rs

```diff
@@ -300,3 +373,102 @@ impl From<bool> for AttrVal {
         AttrVal::Bool(b)
     }
 }
+
+impl From<Nanoseconds> for AttrVal {
+    fn from(ns: Nanoseconds) -> AttrVal {
+        AttrVal::Timestamp(ns)
+    }
+}
+
+impl From<LogicalTime> for AttrVal {
+    fn from(lt: LogicalTime) -> AttrVal {
+        AttrVal::LogicalTime(lt)
+    }
+}
+
+// TODO: Should I actually assume that any UUID is a timeline id?
+impl From<Uuid> for AttrVal {
+    fn from(u: Uuid) -> AttrVal {
+        AttrVal::TimelineId(Box::new(u.into()))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn parse_logical_time() {
+        let reference = Ok(LogicalTime::quaternary(0u64, 0u64, 0u64, 42u64));
+
+        // should parse
+        assert_eq!(reference, "42".parse());
+        assert_eq!(reference, "0:42".parse());
+        assert_eq!(reference, "0:0:42".parse());
+        assert_eq!(reference, "0:0:0:42".parse());
+
+        // should not parse
+        assert_eq!(Err(()), ":".parse::<LogicalTime>());
+        assert_eq!(Err(()), "::".parse::<LogicalTime>());
+        assert_eq!(Err(()), ":0".parse::<LogicalTime>());
+        assert_eq!(Err(()), "0:".parse::<LogicalTime>());
+        assert_eq!(Err(()), "127.0.0.1:8080".parse::<LogicalTime>());
+        assert_eq!(Err(()), "localhost:8080".parse::<LogicalTime>());
+        assert_eq!(Err(()), "example.com:8080".parse::<LogicalTime>());
+    }
+
+    #[test]
+    fn parse_attr_vals() {
+        // Bool
+        assert_eq!(Ok(AttrVal::Bool(false)), "false".parse());
+        assert_eq!(Ok(AttrVal::Bool(true)), "true".parse());
+
+        // Integer
+        assert_eq!(Ok(AttrVal::Integer(37)), "37".parse());
+
+        // BigInt
+        assert_eq!(
+            Ok(BigInt::new_attr_val(36893488147419103232i128)),
+            "36893488147419103232".parse()
+        );
+
+        // Float
+        assert_eq!(Ok(AttrVal::Float(76.37f64.into())), "76.37".parse());
+
+        // TimelineId
+        assert_eq!(
+            Ok(AttrVal::TimelineId(Box::new(
+                Uuid::parse_str("bec14bc0-1dea-4b68-b138-62f7b6827e35")
+                    .unwrap()
+                    .into()
+            ))),
+            "bec14bc0-1dea-4b68-b138-62f7b6827e35".parse()
+        );
+
+        // Timestamp
+        // N.B. This is impossible to parse as an `AttrVal` since it's just a number. Could try
+        // parsing more complex date strings?
+
```

Comment:
I'm assuming Timestamp's inner Nanoseconds is the format we would use in a config file. Is that a bad assumption? Should I try parsing some unambiguous datetime formats to get this? Is there some other format we want to use that wouldn't first be parsed as just an int?

Reply:
Re: Nanoseconds, I think that's probably a good enough assumption.
I don't think we need to add date string parsing now.

Honestly, I think that if we wanted to make more use of the Timestamp variant now, the only approach we have available is to peek over at the relevant `AttrKey` and see if it's one of the designated attributes semantically associated with timey-values. (e.g. `event.timestamp` and `event.interaction.remote_timestamp`).

Reply:
:+1: Okay, I won't change anything here.

> Honestly, I think that if we wanted to make more use of the Timestamp variant now, the only approach we have available is to peek over at the relevant AttrKey and see if it's one of the designated attributes semantically associated with timey-values. (e.g. event.timestamp and event.interaction.remote_timestamp).

Totally, getting a "well known attr key" -> AttrVal type mapping is something that's been on my list to get into the SDK since before it was published.